core/chains/evm/txmgr/finalizer.go

```diff
@@ -0,0 +1,275 @@
+const processHeadTimeout = 10 * time.Minute
+
+type finalizerTxStore interface {
+	FindConfirmedTxes(ctx context.Context, chainID *big.Int) ([]*Tx, error)
+	UpdateTxesFinalized(ctx context.Context, txs []int64, chainId *big.Int) error
+}
+
+type finalizerChainClient interface {
+	BatchCallContext(ctx context.Context, elems []rpc.BatchElem) error
+}
+
+type finalizerHeadTracker interface {
+	LatestAndFinalizedBlock(ctx context.Context) (latest, finalized *evmtypes.Head, err error)
+}
+
+// Finalizer handles processing new finalized blocks and marking transactions as finalized accordingly in the TXM DB
+type evmFinalizer struct {
+	services.StateMachine
+	lggr         logger.SugaredLogger
+	chainId      *big.Int
+	rpcBatchSize int
+
+	txStore     finalizerTxStore
+	client      finalizerChainClient
+	headTracker finalizerHeadTracker
+
+	mb     *mailbox.Mailbox[*evmtypes.Head]
+	stopCh services.StopChan
+	wg     sync.WaitGroup
+}
+
+func NewEvmFinalizer(
+	lggr logger.Logger,
+	chainId *big.Int,
+	rpcBatchSize uint32,
+	txStore finalizerTxStore,
+	client finalizerChainClient,
+	headTracker finalizerHeadTracker,
+) *evmFinalizer {
+	lggr = logger.Named(lggr, "Finalizer")
+	return &evmFinalizer{
+		lggr:         logger.Sugared(lggr),
+		chainId:      chainId,
+		rpcBatchSize: int(rpcBatchSize),
+		txStore:      txStore,
+		client:       client,
+		headTracker:  headTracker,
+		mb:           mailbox.NewSingle[*evmtypes.Head](),
+	}
+}
+
+// Start the finalizer
+func (f *evmFinalizer) Start(ctx context.Context) error {
+	return f.StartOnce("Finalizer", func() error {
+		f.lggr.Debugf("started Finalizer with RPC batch size limit: %d", f.rpcBatchSize)
+		f.stopCh = make(chan struct{})
+		f.wg.Add(1)
+		go f.runLoop()
+		return nil
+	})
+}
+
+// Close the finalizer
+func (f *evmFinalizer) Close() error {
+	return f.StopOnce("Finalizer", func() error {
+		f.lggr.Debug("closing Finalizer")
+		close(f.stopCh)
+		f.wg.Wait()
+		return nil
+	})
+}
+
+func (f *evmFinalizer) Name() string {
+	return f.lggr.Name()
+}
+
+func (f *evmFinalizer) HealthReport() map[string]error {
+	return map[string]error{f.Name(): f.Healthy()}
+}
+
+func (f *evmFinalizer) runLoop() {
+	defer f.wg.Done()
+	ctx, cancel := f.stopCh.NewCtx()
+	defer cancel()
+	for {
+		select {
+		case <-f.mb.Notify():
+			for {
+				if ctx.Err() != nil {
+					return
+				}
+				head, exists := f.mb.Retrieve()
+				if !exists {
+					break
+				}
+				if err := f.ProcessHead(ctx, head); err != nil {
+					f.lggr.Errorw("Error processing head", "err", err)
+					f.SvcErrBuffer.Append(err)
+					continue
+				}
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (f *evmFinalizer) DeliverLatestHead(head *evmtypes.Head) bool {
+	return f.mb.Deliver(head)
+}
+
+func (f *evmFinalizer) ProcessHead(ctx context.Context, head *evmtypes.Head) error {
+	ctx, cancel := context.WithTimeout(ctx, processHeadTimeout)
+	defer cancel()
+	_, latestFinalizedHead, err := f.headTracker.LatestAndFinalizedBlock(ctx)
```

Comment:
Not sure if this has been discussed in the past, but what is the reason for not using the Finalized method on the received head instead?

Reply:
1. At the moment `LatestAndFinalizedBlock` support finality tags even if `FinalityTagBypass = true`. 
2. `LatestAndFinalizedBlock` guarantees to provide latest finalized block. `head.LatestFinalizedHead()` may return nil, if HeadTracker was not fast enough or ws subscription dropped one of the heads.

Reply:
This is a bit concerning if you consider the original design was meant to use `LatestFinalized()` across the board and simplify things. Now we're saying this call is not safe, so every consumer is somewhat forced to import Head Tracker and read directly from its API.  I think this is something that we should fix on the Head Tracker's side and it's currently affecting multiple components.

Reply:
I see entirely different use cases for the callback and `LatestAndFinalizedBlock`:
1. Callback is used when we need to get the latest head as soon as possible to initiate a new iteration (e.g., TXM Confirmer starts looking for receipts, LogPoller starts fetching logs). The head provided by the callback may include the full chain up to the latest finalized block + history depth, but there is no guarantee.
2. `LatestAndFinalizedBlock()` is used when a callback is not convenient and/or the latest finalized block is required. For example, LogPoller decides if it's safe to read logs in batches; Finalizer checks that a tx is finalized; and ChainReader reads a contract's state at the finalized block.

Reply:
If that's the case, then why do we have a `Finalized` method on the head struct if it's not functional or doesn't provide any guarantees? It seems misleading. Also if the goal wasn't to use this head method then we should go ahead and import it on every component and not rely on the callback.